File: android/src/main/kotlin/com/keysking/k_ble_peripheral/delegate/CharacteristicDelegate.kt
package com.keysking.k_ble_peripheral.delegate

import android.bluetooth.BluetoothGattCharacteristic
import java.util.*

object CharacteristicDelegate {
    // 保存所有创建的characteristic,key值是entityId,用于检索
    private val characteristics = mutableMapOf<String, BluetoothGattCharacteristic>()

    fun getEntityId(c: BluetoothGattCharacteristic): String? {
        characteristics.forEach {
            if (it.value == c) {
                return it.key
            }
        }
        return null
    }

    fun createCharacteristic(map: Map<String, Any>): BluetoothGattCharacteristic {
        val uuid = map["uuid"] as String
        val properties = map["properties"] as Int
        val permissions = map["permissions"] as Int
        val entityId = map["entityId"] as String
        val characteristic =
            BluetoothGattCharacteristic(UUID.fromString(uuid), properties, permissions)
        characteristics[entityId] = characteristic
        return characteristic
    }
}

fun BluetoothGattCharacteristic.toMap(): MutableMap<String, Any?> {
    return mutableMapOf(
        Pair("uuid", uuid.toString()),
        Pair("properties", properties),
        Pair("permissions", permissions),
    )
}

File: ios/Classes/CharacteristicDelegate.swift
import CoreBluetooth

class CharacteristicDelegate {
    // Store all created characteristics, using entityId as the key for retrieval
    private var characteristics: [String: KGattCharacteristic] = [:]
    private var requests: [Int: CBATTRequest] = [:]
    private var requestCounter: Int = 0

    func addRequest(request: CBATTRequest) -> Int {
        let requestNumber = requestCounter
        requestCounter = requestCounter + 1

        if(requestCounter == Int.max)
        {
            requestCounter = 0
        }

        requests[requestNumber] = request
        return requestNumber
    }

    func popRequest(requestNumber: Int) -> request?
    {
        guard let request = requests.removeValue(forKey: requestNumber) else {
            return nil
        }
        return request
    }

    func getEntityId(c: CBMutableCharacteristic) -> String? {
        for (key, value) in characteristics {
            if value.characteristic == c {
                return key
            }
        }
        return nil
    }

    func getKChar(entityId: String) -> KGattCharacteristic {
        guard 
            let kChar = characteristics[entityId] 
        else {
            fatalError("Not Found Gatt Characteristic, may be the entityId is wrong.")
        }
        return kChar
    }

    func createCharacteristic(uuid: String, properties: CBCharacteristicProperties, permissions: CBAttributePermissions, entityId: String) -> KGattCharacteristic {
        let characteristic = CBMutableCharacteristic(
            type: CBUUID(string: uuid),
            properties: properties,
            value: nil,
            permissions: permissions
        )

        let kChar = KGattCharacteristic(entityId: entityId, characteristic: characteristic)
        characteristics[entityId] = kChar
        return kChar
    }
}

extension CBMutableCharacteristic {
    func toMap() -> [String: Any?] {
        return [
            "uuid": uuid.uuidString,
            "properties": properties.rawValue,
            "permissions": permissions.rawValue
        ]
    }
}
